Bottom: 57c49bb54c28ffd41da0b641c52c54a95e397ab5
Top:    89e3e265c4ea9adfdf0884d87fe0da1e86b27c12
Author: Ian Kent <raven@themaw.net>
Date:   2013-05-02 14:37:41 +0800

autofs-5.0.7 - use MS_UNBINDABLE


---

diff --git a/daemon/direct.c b/daemon/direct.c
index 228a666..9d8bcc3 100644
--- a/daemon/direct.c
+++ b/daemon/direct.c
@@ -38,6 +38,8 @@
 #define INCLUDE_PENDING_FUNCTIONS
 #include "automount.h"
 
+extern unsigned long mountflags;
+
 /* Attribute to create detached thread */
 extern pthread_attr_t th_attr_detached;
 
@@ -420,7 +422,7 @@ int do_mount_autofs_direct(struct autofs_point *ap,
 
 	map_name = me->mc->map->argv[0];
 
-	ret = mount(map_name, me->key, "autofs", MS_MGC_VAL, mp->options);
+	ret = mount(map_name, me->key, "autofs", mountflags, mp->options);
 	if (ret) {
 		crit(ap->logopt, "failed to mount autofs path %s", me->key);
 		goto out_err;
@@ -768,7 +770,7 @@ int mount_autofs_offset(struct autofs_point *ap, struct mapent *me, const char *
 	if (!type || strcmp(ap->entry->maps->type, "hosts"))
 		map_name = me->mc->map->argv[0];
 
-	ret = mount(map_name, mountpoint, "autofs", MS_MGC_VAL, mp->options);
+	ret = mount(map_name, mountpoint, "autofs", mountflags, mp->options);
 	if (ret) {
 		crit(ap->logopt,
 		     "failed to mount offset trigger %s at %s",
diff --git a/daemon/indirect.c b/daemon/indirect.c
index f4476dc..2dba9bf 100644
--- a/daemon/indirect.c
+++ b/daemon/indirect.c
@@ -37,6 +37,8 @@
 #define INCLUDE_PENDING_FUNCTIONS
 #include "automount.h"
 
+extern unsigned long mountflags;
+
 /* Attribute to create detached thread */
 extern pthread_attr_t th_attr_detached;
 
@@ -146,7 +148,7 @@ static int do_mount_autofs_indirect(struct autofs_point *ap, const char *root)
 	if (!type || strcmp(ap->entry->maps->type, "hosts"))
 		map_name = ap->entry->maps->argv[0];
 
-	ret = mount(map_name, root, "autofs", MS_MGC_VAL, options);
+	ret = mount(map_name, root, "autofs", mountflags, options);
 	if (ret) {
 		crit(ap->logopt,
 		     "failed to mount autofs path %s at %s", ap->path, root);
diff --git a/include/automount.h b/include/automount.h
index e72fa0d..8a78ea0 100644
--- a/include/automount.h
+++ b/include/automount.h
@@ -75,6 +75,10 @@ int load_autofs4_module(void);
 #define SMB_SUPER_MAGIC    0x0000517BL
 #define CIFS_MAGIC_NUMBER  0xFF534D42L
 
+#ifndef MS_UNBINDABLE
+#define MS_UNBINDABLE (1 << 17)
+#endif
+
 /* This sould be enough for at least 20 host aliases */
 #define HOST_ENT_BUF_SIZE	2048
 
diff --git a/lib/mounts.c b/lib/mounts.c
index 550445c..9f76623 100644
--- a/lib/mounts.c
+++ b/lib/mounts.c
@@ -17,8 +17,8 @@
 #include <limits.h>
 #include <sys/types.h>
 #include <sys/stat.h>
-#include <sys/ioctl.h>
 #include <sys/mount.h>
+#include <sys/ioctl.h>
 #include <sys/wait.h>
 #include <ctype.h>
 #include <stdio.h>
@@ -46,6 +46,8 @@ static const char mnt_name_template[]      = "automount(pid%u)";
 static struct kernel_mod_version kver = {0, 0};
 static const char kver_options_template[]  = "fd=%d,pgrp=%u,minproto=3,maxproto=5";
 
+unsigned long mountflags = MS_UNBINDABLE|MS_MGC_VAL;
+
 unsigned int linux_version_code(void)
 {
 	struct utsname my_utsname;
@@ -102,11 +104,14 @@ unsigned int query_kproto_ver(void)
 		return 0;
 	}
 
-	if (mount("automount", t_dir, "autofs", MS_MGC_VAL, options)) {
-		close(pipefd[0]);
-		close(pipefd[1]);
-		rmdir(t_dir);
-		return 0;
+	if (mount("automount", t_dir, "autofs", mountflags, options)) {
+		mountflags &= ~MS_UNBINDABLE;
+		if (mount("automount", t_dir, "autofs", mountflags, options)) {
+			close(pipefd[0]);
+			close(pipefd[1]);
+			rmdir(t_dir);
+			return 0;
+		}
 	}
 
 	close(pipefd[1]);
diff --git a/modules/parse_sun.c b/modules/parse_sun.c
index ae1caf7..d75e117 100644
--- a/modules/parse_sun.c
+++ b/modules/parse_sun.c
@@ -29,7 +29,6 @@
 #include <sys/vfs.h>
 #include <sys/utsname.h>
 #include <netinet/in.h>
-#include <sys/mount.h>
 #include <linux/fs.h>
 
 #define MODULE_PARSE
